recipes/full_finetune_distributed.py

```diff
@@ -468,44 +477,47 @@ def train(self) -> None:
                 # Compute loss
                 loss = self._loss_fn(logits, labels)
 
-                # Note: We're always logging the loss before normalizing it
-                # Check if this is the norm or not
-                if (
-                    self.total_training_steps % self._log_every_n_steps == 0
-                    and self._is_rank_zero
-                ):
-                    pbar.set_description(f"{curr_epoch+1}|{idx+1}|Loss: {loss.item()}")
-                    self._metric_logger.log_dict(
-                        {
-                            "loss": loss.item(),
-                            "lr": self._optimizer.param_groups[0]["lr"],
-                            "gpu_mem_alloc": torch.cuda.memory_allocated() / 1e9,
-                            "gpu_mem_reserved": torch.cuda.memory_reserved() / 1e9,
-                        },
-                        step=self.total_training_steps,
-                    )
-
                 loss = loss / self._gradient_accumulation_steps
+                running_loss += loss
                 loss.backward()
 
+                # Step with optimizer
                 if (idx + 1) % self._gradient_accumulation_steps == 0:
                     self._optimizer.step()
                     self._optimizer.zero_grad(set_to_none=True)
 
                     # Update the number of steps when the weights are updated
                     self.total_training_steps += 1
 
-                # Log peak memory for iteration
-                if (
-                    self.total_training_steps % self._log_peak_memory_every_n_steps == 0
-                    and self._is_rank_zero
-                ):
-                    # Log peak memory for iteration
-                    memory_stats = utils.get_memory_stats(device=self._device)
-                    self._metric_logger.log_dict(
-                        memory_stats, step=self.total_training_steps
+                    loss_to_log = running_loss.item()
+                    pbar.update(1)
+                    pbar.set_description(
+                        f"{curr_epoch+1}|{self.total_training_steps+1}|Loss: {loss_to_log}"
                     )
 
+                    # Log per-step metrics
+                    if (
+                        self.total_training_steps % self._log_every_n_steps == 0
+                        and self._is_rank_zero
+                    ):
+                        time_per_step = time.perf_counter() - t0
+                        log_dict = {
+                            "loss": loss_to_log,
+                            "lr": self._optimizer.param_groups[0]["lr"],
+                            "tokens_per_second": num_tokens / time_per_step,
```

Comment:
Yeah this is a good point. Actually I'm not sure what the right thing to do here is for two reasons. (1) We only log rank-zero loss, and (2) I think a naive multiplication with num_devices will not work. If I understand correctly DistributedSampler will apply collate_fn to each worker's batch separately, in which case we actually have different numbers of tokens on each rank, right?

Reply:
Yeah the more I think about it the less I want to change this. (1) it will introduce unneeded syncs to calculate this value properly and (2) we could just log on every rank if we really want to and aggregate posthoc